Introductory documentation on Callbacks and fixed some mutability oversights

aurorameccanicadev opened the pull request:
(refers to #10)
 
I made the following changes:
- added an introductory example in the main documentation page
- added example for callbacks
- some setter methods were accepting `&self `instead of `&mut self`, hence not enforcing the rust mutability rules
- there were leftovers of some async methods. However, while the method was commented out, its docstring wasn't, so it was leaking into the documentation of the next method

fpagliughi commented:
> some setter methods were accepting &self instead of &mut self, hence not enforcing the rust mutability rules

The C library is thread safe, and thus the objects can be thought of as having interior mutability from the Rust perspective. The `&mut self` is not needed.

fpagliughi commented:
Just getting back to some open-source stuff now that vacation and other travel has completed.

I would gladly add the documentation updates, but the _mut_ additions are not desirable. Unfortunately there isn't a clean separations of commits so that I could even cherry-pick the documentation updates.

In general it is preferable to make separate pull requests for different things. That way, different features can be discussed and merged independently.

aurorameccanicadev commented:
> Just getting back to some open-source stuff now that vacation and other travel has completed.
> 
> I would gladly add the documentation updates, but the _mut_ additions are not desirable. Unfortunately there isn't a clean separations of commits so that I could even cherry-pick the documentation updates.
> 
> In general it is preferable to make separate pull requests for different things. That way, different features can be discussed and merged independently.

Ok, I'll delete the mutability changes and include only the docs improvements, I'll do it probably in the next week when I have some free time.  Thank you for your review :)

aurorameccanicadev commented:
I forced a commit to my fork so I deleted my previous wrong commits.
Now this pull request only has one commit containing the doc changes.
The commit message contains more information on what I changed

Let me know if I can improve it in some way

fpagliughi commented:
No. Cool. Thanks for the help!
